### PR TITLE
Prometheus: Respect dashboard queries when querying ad hoc filter labels

### LINKED ISSUE #84276: Prometheus: Respect dashboard queries when querying ad hoc filter labels
Currently, when presenting the ad hoc filters, prometheus data source shows all available labels in the labels selector. 

With https://github.com/grafana/grafana/pull/81071 merged it's now possible for the ad hoc filters to access the queries that exist in the Grafana dashboard. This allows data source to narrow down the list of presented ad hoc filter labels to a list that' meaningful in the context of a given dashboard.

However, no data source currently does that, and Prometheus seems like an obvious candidate to introduce this feature. 

Please have a look at  https://github.com/grafana/grafana/pull/81071  to see the API changes that enable this feature.

### PR DESCRIPTION
**Pull request by itsmylife**
**What is this feature?**

When querying ad-hoc filter, we can now access the dashboard queries. So we can use those queries to filter out unnecessary keys that we cannot apply to current metrics. But this works only with the new architecture. For that `dashboardScene` feature flag must be enabled. 

**Why do we need this feature?**

Better ad-hoc filter queries

**Who is this feature for?**

Prometheus users

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/84276

**Special notes for your reviewer:**
### How to test
- Simply create a Prometheus dashboard and add an ad-hoc filter template variable
- Create a couple of panels that have different metrics
- Click to add ad-hoc filter and see all the labels are fetched
- Switch to this branch and try again. 
- See only applicable labels are fetched.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-prometheus/jest.config.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+const sharedConfig = require('../../jest.config.js');
+module.exports = {
+  ...sharedConfig,
+  rootDir: '../../',
+};
```
Comment (itsmylife) [inline packages/grafana-prometheus/jest.config.js:5]: This is not related to the PR but necessary for testing with Intellij IDEs.
image

**Review by NWRichmond — COMMENTED** [review]
I might be misunderstanding something, but on `ismail/adhoc-query-filters` (and with `usePrometheusFrontendPackage` enabled), I'm seeing label names in ad-hoc queries that aren't related to the metrics in my dashboard panels.

I have two panels which are querying different metrics that share the same label names (`instance` and `job`):

In-Use Bytes label names

goroutines label names

In my dashboard, the ad-hoc query is showing label names that are unrelated to the metrics I'm querying in these panels:

https://github.com/grafana/grafana/assets/5732000/b43d050c-a18b-4229-8de9-a992648da141

Based on the **How to test** steps and linked issue, my expectation is that only `instance` and `job` would appear in the label name part of the ad-hoc query.

I'm also curious if you are planning to implement this feature in `public/app/plugins/datasource/prometheus` too?

**Comment by kylebrandt** [comment]
Two things:
 - A) Ad-hoc filters should be limited by the metrics in the queries. So we will add a resource endpoint that is kind of like labels I think, (we can do this _after_ item B, and use what is in this PR for now)
 - B) (This one first, in this PR) We want the BE parse to handle applying scopes and ad-hoc filters. So the ad-hoc filter should be send as a parameter of the query alongside the scope (waiting on prop name) and then the BE will apply the filters and scopes to the expr property.

**Comment by itsmylife** [comment]
@NWRichmond `dashboardScene` feature toggle must be enabled too. This feature depends on that feature toggle.

**Review by NWRichmond — COMMENTED** [review]
Thanks, @itsmylife! And sorry for missing that feature toggle note in the PR description.

After enabling `dashboardScene`, only the relevant label names appear:

https://github.com/grafana/grafana/assets/5732000/6c144f5c-6f74-4ae7-8549-14dbcbdc3b0a

Nice!

I'll be happy to verify that everything is still working after backend updates are made.

**Comment by itsmylife** [comment]
@NWRichmond Backend updates are not related to this feature. This is only for filling the dropdown.

**Review by NWRichmond — APPROVED** [review]
Thanks, @itsmylife. In that case, ✅

**Pull request merged**